Add test for restore adjust and re-restore

KillianG opened the pull request:
add a test for reducing restore duration and a test for increasing restore duration

Issue: ZENKO-4727

bert-e commented:
# Hello killiang,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/zenko/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ZENKO-4727 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `2.8.23`

* `2.9.0`


Please check the `Fix Version/s` of ZENKO-4727, or the target
branch of this pull request.

francoisferrand commented on tests/ctst/features/azureArchive.feature:
```diff
@@ -121,6 +121,38 @@ Feature: Azure Archive
+        | versioningConfiguration | objectCount | objectSize | restoreDays |
+        |           Non versioned |           2 |        100 |           15 |
+        |               Versioned |           2 |        100 |           15 |
+        |               Suspended |           2 |        100 |           15 |
```
as per the ticket, should also add a test case `restore→expiration→re-restore` (and verify it expires a second time as expected)

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ZENKO-4727 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `2.8.24`

* `2.9.0`


Please check the `Fix Version/s` of ZENKO-4727, or the target
branch of this pull request.

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

francoisferrand commented on tests/ctst/common/common.ts:
```diff
@@ -258,18 +271,14 @@ Given('a transition workflow to {string} location', async function (this: Zenko,
+    await restoreObject.call(this, days, objectName);
 });
 
+When('i adjust restore duration of object {string} for {int} days',
```
"adjust restore" is not a separate step : this is not intrinsically an adjust, but only because of the condition of the object when this happens...

→ the steps should represent what we actually do (e.g. "issue an s3:restore command"), to make it easier to understand what goes on. I.e. we may have a "adjust restore" **_scenario_**, but that scenario would be simply "archive → restore → wait until restored → restore → check the restore duration has been updated"

francoisferrand commented on tests/ctst/steps/azureArchive.ts:
```diff
@@ -373,7 +373,7 @@ When('i run sorbetctl to retry failed restore for {string} location', async func
     const realTimeDay = days * 24 * 60 * 60 * 1000 /
         (this.parameters.TimeProgressionFactor > 1 ? this.parameters.TimeProgressionFactor : 1);
-    await Utils.sleep(realTimeDay + 200 * 1000); // We also wait for sorbet forwarder to see the object has expired
+    await Utils.sleep(realTimeDay); // We also wait for sorbet forwarder to see the object has expired
```
seems unrelated : if needed, should go in a separate commit to explain what happens here...

(and comment should probably be updated, since we don't have the "extra' delay?)

Commit:
Add test scenario for restore adjust

Add 2 ctst test scenarios in which we test the restore of an already restored object and the restore operation on a restore-expired object

Issue: ZENKO-4727

Commit:
Remove extra wait in ctst scenario

Remove the extra wait when waiting for the restore to expire, 200k was overkill and we don't need it anymore

Issue: ZENKO-4727

welansari commented on tests/ctst/features/iam-policies/AssumeRoleWithWebIdentity.feature:
```diff
@@ -125,52 +125,52 @@ Feature: Assume Role with Web Identity
+      | GetObjectLockConfiguration      | STORAGE_MANAGER       | without        | exists         | ObjectLockConfigurationNotFoundError            |
+      | GetObjectRetention              | STORAGE_MANAGER       | with           | exists         | Bucket is missing Object Lock Configuration     |
+      | GetObjectLegalHold              | STORAGE_MANAGER       | with           | exists         | Bucket is missing Object Lock Configuration     |
+      | PutObjectRetention              | STORAGE_MANAGER       | with           | exists         | MalformedXML: request xml is undefined or empty |
```
nit: for consistency with the other expected errors, should we only specify the error type `MalformedXML` without the description ?

welansari commented on tests/ctst/features/azureArchive.feature:
```diff
@@ -121,6 +121,70 @@ Feature: Azure Archive
+    @Flaky
+    @AzureArchive
+    @ColdStorage
+    @WIP
```
leftover from debugging ?

welansari commented on tests/ctst/features/azureArchive.feature:
```diff
@@ -121,6 +121,70 @@ Feature: Azure Archive
+    @Flaky
+    @AzureArchive
+    @ColdStorage
+    @WIP
```
same here

francoisferrand commented on solution/deps.yaml:
```diff
@@ -16,7 +16,7 @@ cloudserver:
   dashboard: cloudserver-dashboards
   image: cloudserver
-  tag: 8.8.7
+  tag: 8.8.20
```
already in #2023 I guess : will need to rebase after it is merged

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ZENKO-4727 contains:


* `2.8.26`

* `2.9.0`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `2.8.27`

* `2.9.0`


Please check the `Fix Version/s` of ZENKO-4727, or the target
branch of this pull request.

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

KillianG commented:
/create_integration_branches

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/ZENKO-4727-add-missing-test-for-restore-duration` into
`development/2.8`
* [w/2.9/improvement/ZENKO-4727-add-missing-test-for-restore-duration](https://github.com/scality/zenko/compare/development/2.9...w/2.9/improvement/ZENKO-4727-add-missing-test-for-restore-duration) will be merged into `development/2.9` 



The following branches will **NOT** be impacted:

* `development/2.5`
* `development/2.6`
* `development/2.7`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **create_integration_branches**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**

Commit:
Kafka cleaner test more robust

Issue: ZENKO-4727

Commit:
Fix KFKC test flakyness and bump ZKOP

Issue: ZENKO-4727

francoisferrand commented:
/wait

francoisferrand commented on solution/deps.yaml:
```diff
@@ -99,7 +99,7 @@ vault:
   sourceRegistry: registry.scality.com/zenko-operator
   image: zenko-operator
-  tag: 1.5.43
+  tag: 1.5.44
```
not needed anymore, already bump'ed

francoisferrand commented on tests/ctst/common/common.ts:
```diff
@@ -300,40 +300,58 @@ Then('i {string} be able to add user metadata to object {string}',
     });
 
-Then('kafka consumed messages should not take too much place on disk',
+Then('kafka consumed messages should not take too much place on disk', { timeout: 130000 },
```
timeout should should be based on `KafkaCleanerInterval`, i.e. a number of time the interval?

KillianG replied:
Should be great but unfortuneatly we can't access `Zenko` object here :(
And I am not sure if this is great anyway, because if we set kafkacleaner interval to 10mn for some reason, we'll end up with 30mn timeout ?

francoisferrand commented on tests/ctst/common/common.ts:
```diff
@@ -300,40 +300,58 @@ Then('i {string} be able to add user metadata to object {string}',
-        
-        // 10 seconds added to be sure kafkacleaner had time to process
-        await Utils.sleep(seconds * 1000 + 10000);
+        while (topics.length > 0 && new Date().getTime() - timeStart < 120000) {
```
what is the point of checking the duration?
* the timeout on the test is designed to catch test taking too long to complete
* here we simply want to try "a few times" (or even not limit, and just fail without timeout if it never settles...)

KillianG replied:
This is to end properly and not getting killed by the timeout so at the end we can log each topic that is not cleaned in time, if we just wait for the timeout, we'll end up with no readable logs.. just "test timed out" with no info..

francoisferrand commented on tests/ctst/common/common.ts:
```diff
@@ -300,40 +300,58 @@ Then('i {string} be able to add user metadata to object {string}',
+            const previousOffsets = await getTopicsOffsets(topics, kafkaAdmin);
 
-        const newOffsets = await getTopicsOffsets(topics, kafkaAdmin);
+            const seconds = parseInt(this.parameters.KafkaCleanerInterval);
```
no point doing this in the loop, just compute once before the loop

francoisferrand commented on tests/ctst/common/common.ts:
```diff
@@ -300,40 +300,58 @@ Then('i {string} be able to add user metadata to object {string}',
             }
         }
+
+        assert(topics.length === 0, `Topics ${topics.join(', ')} still have not been cleaned`);
```
here we should fail if we exited the loop with the "`newMessagesAfterClean`" case, i.e. if it never converged

KillianG replied:
Maybe i missed something, but this is already what it does, while there is `newMessagesAfterClean` we continue (i.e we don't remove the topic from the list of topics to be checked) and here, if there is topics that remains in the array, it's that they haven't been cleaned and we print them to be able to debug in case it happens

francoisferrand commented on tests/ctst/common/common.ts:
```diff
@@ -300,40 +300,58 @@ Then('i {string} be able to add user metadata to object {string}',
+            for (let i = 0; i < topics.length; i++) {
+                process.stdout.write(`\nChecking topic ${topics[i]}\n`);
+                for (let j = 0; j < newOffsets[i].partitions.length; j++) {
+                    // Checking that the min offset has increased due to kafkacleaner
+                    // or that it didn't need to change because there was no new messages
+                    // or new messages came after a clean but still cleaned to previous high offset
```
the comment is hard to relate with the code now...

francoisferrand commented on tests/ctst/common/common.ts:
```diff
@@ -300,40 +300,58 @@ Then('i {string} be able to add user metadata to object {string}',
+                    // Checking that the min offset has increased due to kafkacleaner
+                    // or that it didn't need to change because there was no new messages
+                    // or new messages came after a clean but still cleaned to previous high offset
+                    const lowOffsetIncreased = newOffsets[i].partitions[j].low > previousOffsets[i].partitions[j].low;
+                    const noNewMessages = newOffsets[i].partitions[j].high === newOffsets[i].partitions[j].low;
```
not used here → move after the `if` for clarify, where they will be used

Commit:
Refactor code to improve performance and readability

Issue: ZENKO-4727

Commit:
Add timeout based on KFKC interval instead of hardcode it

Issue: ZENKO-4727

Commit:
Clear timeout before leaving KFKC test

Issue: ZENKO-4727

francoisferrand commented on tests/ctst/common/common.ts:
```diff
@@ -300,40 +300,68 @@ Then('i {string} be able to add user metadata to object {string}',
-                previousOffsets: ${previousOffsets[i].partitions[j].low} / ${previousOffsets[i].partitions[j].high},
-                newOffsets: ${newOffsets[i].partitions[j].low} / ${newOffsets[i].partitions[j].high}`);
+
+
```
nit: extra empty line

francoisferrand commented on tests/ctst/common/common.ts:
```diff
@@ -300,40 +300,68 @@ Then('i {string} be able to add user metadata to object {string}',
+                }
             }
         }
+        clearTimeout(timeoutID);
```
in case of `assert` before, this will not be executed : which could cause strange behavior.

a safer approach would be to use `try{}finally{}` block:

```suggestion
        const timeoutID = setTimeout(....);
        try {
           //run the check
        } finally {
           clearTimeout(timeoutID);
        }
```

(atlernatively, we could maybe [`async.timeout`](https://caolan.github.io/async/v3/docs.html#timeout) instead)

Commit:
Add try finally to be sure cleanTimeout is called

Issue: ZENKO-4727

francoisferrand commented on tests/ctst/common/common.ts:
```diff
@@ -300,40 +300,72 @@ Then('i {string} be able to add user metadata to object {string}',
-                previousOffsets: ${previousOffsets[i].partitions[j].low} / ${previousOffsets[i].partitions[j].high},
-                newOffsets: ${newOffsets[i].partitions[j].low} / ${newOffsets[i].partitions[j].high}`);
+
+        try {
```
should encompass listing of kafka topic as well, i.e. start immediately after creating the timer

Commit:
Refactor Kafka topic cleanup logic in common.ts

Issue: ZENKO-4727

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ZENKO-4727 contains:


* `2.8.28`

* `2.9.1`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `2.8.29`

* `2.9.2`


Please check the `Fix Version/s` of ZENKO-4727, or the target
branch of this pull request.




*The following options are set:* **create_integration_branches**

bert-e commented:
# History mismatch



Merge commit #e487dd11594c1832187ee7f6704b0e0a233e16a9 on the integration branch
`w/2.9/improvement/ZENKO-4727-add-missing-test-for-restore-duration` is merging a branch which is neither the current
branch `improvement/ZENKO-4727-add-missing-test-for-restore-duration` nor the development branch
`development/2.9`.

It is likely due to a rebase of the branch `improvement/ZENKO-4727-add-missing-test-for-restore-duration` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **create_integration_branches**

KillianG commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **create_integration_branches**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/ZENKO-4727-add-missing-test-for-restore-duration` into
`development/2.8`
* [w/2.9/improvement/ZENKO-4727-add-missing-test-for-restore-duration](https://github.com/scality/zenko/compare/development/2.9...w/2.9/improvement/ZENKO-4727-add-missing-test-for-restore-duration) will be merged into `development/2.9` 



The following branches will **NOT** be impacted:

* `development/2.5`
* `development/2.6`
* `development/2.7`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **create_integration_branches**

KillianG commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/Zenko/commit/a5afc48edb0b7ed8ca23edac4f4e023496cc88d9) did not succeed in branch w/2.9/improvement/ZENKO-4727-add-missing-test-for-restore-duration.




*The following options are set:* **approve, create_integration_branches**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.8`

* :heavy_check_mark: `development/2.9`



The following branches will **NOT** be impacted:

* `development/2.5`
* `development/2.6`
* `development/2.7`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve, create_integration_branches**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.8`

* :heavy_check_mark: `development/2.9`



The following branches have **NOT** changed:

* `development/2.5`
* `development/2.6`
* `development/2.7`



Please check the status of the associated issue ZENKO-4727.

Goodbye killiang.
